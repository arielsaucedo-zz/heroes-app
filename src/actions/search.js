import axios from "axios";
import { types } from "../types/types";

const baseUrl = process.env.REACT_APP_API_URL;

export const startSearch = (searchtext) => {
  return async (dispatch) => {
    const resp = await axios.get(`${baseUrl}/search/${searchtext}`, {});
    if (resp.data.response === "success") {
      dispatch(search(resp.data.results));
    } else {
      dispatch(clearSearch());
    }
  };
};

const search = (results) => {
  return {
    type: types.searchStart,
    payload: results,
  };
};

const clearSearch = () => {
  return {
    type: types.clearSearch,
  };
};
