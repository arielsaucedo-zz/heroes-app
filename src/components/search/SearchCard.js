import React from "react";
import { useDispatch, useSelector } from "react-redux";
import Swal from "sweetalert2";

import { addNewBadHero, addNewGoodHero } from "../../actions/hero";

export const SearchCard = ({ id, name, image }) => {
  const dispatch = useDispatch();
  const { results } = useSelector((state) => state.search);
  const state = useSelector((state) => state.team);

  const handleHeroClick = () => {
    const newHero = results.filter((hero) => hero.id === id);
    const { biography } = newHero[0];

    if (biography.alignment === "good") {
      const heroInTeam = state.good.filter((hero) => hero.id === id);

      if (heroInTeam[0] !== newHero[0] && state.good.length < 3) {
        dispatch(addNewGoodHero(newHero));
      } else {
        Swal.fire("No es posible sumar a este heroe a tu equipo");
      }
    } else {
      const heroInTeam = state.bad.filter((hero) => hero.id === id);

      if (heroInTeam[0] !== newHero[0] && state.bad.length < 3) {
        dispatch(addNewBadHero(newHero));
      } else {
        Swal.fire("No es posible sumar a este heroe a tu equipo");
      }
    }
  };

  return (
    <div className="card d-flex flex-column rounded" style={{ maxWidth: 200 }}>
      <div>
        <img src={image.url} alt={name} className="img-thumbnail img-fluid" />
      </div>
      <div className="m-2 text-dark d-flex flex-column justify-content-around">
        <h4>{name}</h4>
        <br />
        <div className="text-center">
          <div className="mt-3 d-flex justify-content-around">
            <button className="btn btn-success w-100" onClick={handleHeroClick}>
              Agregar
            </button>
          </div>
        </div>
      </div>
    </div>
  );
};
