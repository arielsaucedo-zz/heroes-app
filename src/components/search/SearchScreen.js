import React from "react";
import { useDispatch, useSelector } from "react-redux";
import { Formik } from "formik";

import { startSearch } from "../../actions/search";
import { SearchCard } from "./SearchCard";

export const SearchScreen = () => {
  const dispatch = useDispatch();

  const { results, search } = useSelector((state) => state.search);
  console.log(results);

  return (
    <div className="d-flex justify-content-center ">
      <div>
        <div>
          <h2>Search Form</h2>
          <hr />
          <Formik
            initialValues={{ searchText: "" }}
            validate={(values) => {
              const errors = {};
              if (values.searchText === "") {
                errors.searchText =
                  "Este campo es requerido para realizar la busqueda";
              }
              return errors;
            }}
            onSubmit={(values) => {
              return dispatch(startSearch(values.searchText)).catch((err) => {
                console.log(err);
              });
            }}
          >
            {({
              values,
              touched,
              errors,
              handleChange,
              handleSubmit,
              isSubmitting,
            }) => (
              <form
                onSubmit={handleSubmit}
                className="d-flex flex-column justify-content-center"
              >
                <input
                  type="text"
                  placeholder="Find your hero"
                  className="form-control w-50 align-self-center"
                  name="searchText"
                  autoComplete="off"
                  value={values.searchText}
                  onChange={handleChange}
                />
                <span>
                  {errors.searchText && touched.searchText && errors.searchText}
                </span>
                <button
                  type="submit"
                  className="btn btn-primary mt-3 w-50 align-self-center"
                  disabled={isSubmitting}
                >
                  Search
                </button>
              </form>
            )}
          </Formik>
        </div>
        <div className="d-flex row justify-content-center mt-4 w-100">
          <h2 className="text-center">Resultados</h2>
          {search === null && (
            <h3 className="text-center w-100">Busca tu heroe para tu equipo</h3>
          )}
          {search === "ok" && results.length === 0 && (
            <div className="alert alert-danger text-center fw-bolder fs-4">
              No existe ningun heroe para su busqueda
            </div>
          )}

          <div className="d-flex justify-content-evenly flex-wrap">
            {results.map((hero) => (
              <div className="m-1" key={hero.id}>
                <SearchCard {...hero} />
              </div>
            ))}
          </div>
        </div>
      </div>
    </div>
  );
};
