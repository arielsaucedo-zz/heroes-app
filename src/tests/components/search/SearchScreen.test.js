import { mount } from "enzyme";
import { Provider } from "react-redux";
import { fireEvent, render, waitFor } from "@testing-library/react";

import configureStore from "redux-mock-store";
import thunk from "redux-thunk";

import { SearchScreen } from "../../../components/search/SearchScreen";

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initialState = {
  search: {
    search: null,
    results: [],
  },
};

let store = mockStore(initialState);
store.dispatch = jest.fn();

const wrapper = mount(
  <Provider store={store}>
    <SearchScreen />
  </Provider>
);

describe("Pruebas en componente <SearchScreen />", () => {
  beforeEach(() => {
    store = mockStore(initialState);
    jest.clearAllMocks();
  });

  test("Debe de mostrarse correctamente", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("Debe de mostrar mensaje si al buscar API no responde", async () => {
    const initState = {
      search: {
        search: "ok",
        results: [],
      },
    };

    const store = mockStore(initState);

    const wrapper = mount(
      <Provider store={store}>
        <SearchScreen />
      </Provider>
    );

    expect(wrapper.find(".alert").exists()).toBe(true);
    expect(wrapper.find(".alert").text().trim()).toEqual(
      "No existe ningun heroe para su busqueda"
    );
  });
});
