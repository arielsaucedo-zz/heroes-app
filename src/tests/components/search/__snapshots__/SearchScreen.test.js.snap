// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en componente <SearchScreen /> Debe de mostrarse correctamente 1`] = `
<div
  className="d-flex justify-content-center "
>
  <div>
    <div>
      <h2>
        Buscador
      </h2>
      <hr />
      <form
        className="d-flex flex-column justify-content-center"
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control w-50 align-self-center"
          data-testid="searchText"
          name="searchText"
          onChange={[Function]}
          placeholder="Busca a tu heroe"
          type="text"
          value=""
        />
        <span
          className="text-center mt-1"
        />
        <button
          className="btn btn-primary mt-3 w-50 align-self-center"
          data-testid="button"
          disabled={false}
          type="submit"
        >
          Buscar
        </button>
      </form>
    </div>
    <div
      className="d-flex row justify-content-center mt-4 w-100"
    >
      <h2
        className="text-center"
      >
        Resultados
      </h2>
      <h3
        className="text-center w-100"
      >
        Busca tu heroe para tu equipo
      </h3>
      <div
        className="d-flex justify-content-evenly flex-wrap"
      />
    </div>
  </div>
</div>
`;
